### PR TITLE
Importer GUI redesign

### PR DESCRIPTION
**Pull request by abdelr**

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -21,6 +21,7 @@ skip_branch_with_pr: true
 branches:
   only:
     - develop
+    - feature/relative_expression_redesign
```
Comment (msevestre) [inline appveyor.yml:24]: remove

File: src/PKSim.Assets.Images/PKSim - Backup.Assets.Images.csproj [hunk 2]
```diff
@@ -0,0 +1,56 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (msevestre) [inline src/PKSim.Assets.Images/PKSim - Backup.Assets.Images.csproj:1]: Remove file

File: src/PKSim.Assets/PKSim - Backup.Assets.csproj [hunk 3]
```diff
@@ -0,0 +1,39 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (msevestre) [inline src/PKSim.Assets/PKSim - Backup.Assets.csproj:1]: Remove file

File: src/PKSim.CLI.Core/MinimalImplementations/CLIDataImporter.cs [hunk 4]
```diff
@@ -12,9 +14,15 @@ public IEnumerable<DataRepository> ImportDataSets(IReadOnlyList<MetaDataCategory
...
+      public IEnumerable<DataRepository> ImportFromConfiguration(ImporterConfiguration configuration, bool promptForConfirmation, IReadOnlyList<MetaDataCategory> metaDataCategories,
+         IReadOnlyList<ColumnInfo> columnInfos, DataImporterSettings dataImporterSettings)
       {
-         return new NullDataRepository();
+         return Enumerable.Empty<DataRepository>();
+      }
+
+      (IEnumerable<DataRepository> DataRepositories, ImporterConfiguration Configuration) IDataImporter.ImportDataSets(IReadOnlyList<MetaDataCategory> metaDataCategories, IReadOnlyList<ColumnInfo> columnInfos, DataImporterSettings dataImporterSettings)
+      {
+         return (new List<DataRepository>(), null);
```
Comment (msevestre) [inline src/PKSim.CLI.Core/MinimalImplementations/CLIDataImporter.cs:25]: use return Enumerable.Empty<DataRepository>() as in the line above for consistency

File: src/PKSim.CLI.Core/MinimalImplementations/CLIDataImporter.cs [hunk 5]
```diff
@@ -12,9 +14,15 @@ public IEnumerable<DataRepository> ImportDataSets(IReadOnlyList<MetaDataCategory
...
 
-      public DataRepository ImportDataSet(IReadOnlyList<MetaDataCategory> metaDataCategories, IReadOnlyList<ColumnInfo> columnInfos, DataImporterSettings dataImporterSettings)
+      public IEnumerable<DataRepository> ImportFromConfiguration(ImporterConfiguration configuration, bool promptForConfirmation, IReadOnlyList<MetaDataCategory> metaDataCategories,
+         IReadOnlyList<ColumnInfo> columnInfos, DataImporterSettings dataImporterSettings)
       {
-         return new NullDataRepository();
+         return Enumerable.Empty<DataRepository>();
+      }
+
+      (IEnumerable<DataRepository> DataRepositories, ImporterConfiguration Configuration) IDataImporter.ImportDataSets(IReadOnlyList<MetaDataCategory> metaDataCategories, IReadOnlyList<ColumnInfo> columnInfos, DataImporterSettings dataImporterSettings)
```
Comment (msevestre) [inline src/PKSim.CLI.Core/MinimalImplementations/CLIDataImporter.cs:23]: why do you need explicit implementation here?`IDataImporter.`

File: src/PKSim.Core/PKSim - Backup.Core.csproj [hunk 6]
```diff
@@ -0,0 +1,51 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (msevestre) [inline src/PKSim.Core/PKSim - Backup.Core.csproj:1]: Remove

File: src/PKSim.Core/Services/IImportObservedDataTask.cs [hunk 7]
```diff
@@ -1,10 +1,14 @@
...
+﻿using OSPSuite.Core.Import;
+using PKSim.Core.Model;
 
 namespace PKSim.Core.Services
 {
    public interface IImportObservedDataTask
    {
       void AddObservedDataToProject();
       void AddObservedDataToProjectForCompound(Compound compound);
+      void AddObservedDataFromXmlToProjectForCompound(Compound subject);
```
Comment (msevestre) [inline src/PKSim.Core/Services/IImportObservedDataTask.cs:10]: I don't understand the `FromXml` signature.  What use case is this?
Rename subject to compound

File: src/PKSim.Core/Services/IImportObservedDataTask.cs [hunk 8]
```diff
@@ -1,10 +1,14 @@
...
+      void AddObservedDataFromConfigurationToProject(ImporterConfiguration configuration);
```
Comment (msevestre) [inline src/PKSim.Core/Services/IImportObservedDataTask.cs:11]: Don't we need the compound for this one?

File: src/PKSim.Infrastructure/Reporting/TEX/Reporters/ProjectReporter.cs [hunk 9]
```diff
@@ -35,6 +35,7 @@ public override IReadOnlyCollection<object> Report(PKSimProject project, OSPSuit
          list.AddRange(_simulationsReporter.Report(project.All<Simulation>().ToList(), tracker));
          list.AddRange(_simulationComparisonsReporter.Report(project.AllSimulationComparisons.ToList(), tracker));
          list.AddRange(_observedDataReporter.Report(project.AllObservedData.ToList(), tracker));
+         list.AddRange(_observedDataReporter.Report(project.AllImporterConfigurations.ToList(), tracker));
```
Comment (msevestre) [inline src/PKSim.Infrastructure/Reporting/TEX/Reporters/ProjectReporter.cs:38]: I don't think this makes sense to export this info to PDF. 
I would remove it for now (or you also need to create the Exporter for PDF for a configuration)

**Review by msevestre — CHANGES_REQUESTED** [review]

File: appveyor.yml (hunk 1 above)
Comment (abdelr) [reply appveyor.yml:24]: Done

File: src/PKSim.Assets.Images/PKSim - Backup.Assets.Images.csproj (hunk 2 above)
Comment (abdelr) [reply src/PKSim.Assets.Images/PKSim - Backup.Assets.Images.csproj:1]: Done

File: src/PKSim.Assets/PKSim - Backup.Assets.csproj (hunk 3 above)
Comment (abdelr) [reply src/PKSim.Assets/PKSim - Backup.Assets.csproj:1]: Done

File: src/PKSim.CLI.Core/MinimalImplementations/CLIDataImporter.cs (hunk 4 above)
Comment (abdelr) [reply src/PKSim.CLI.Core/MinimalImplementations/CLIDataImporter.cs:25]: Done

File: src/PKSim.CLI.Core/MinimalImplementations/CLIDataImporter.cs (hunk 5 above)
Comment (abdelr) [reply src/PKSim.CLI.Core/MinimalImplementations/CLIDataImporter.cs:23]: We are keeping the same logic as before (with the old importer), this is a dependency of other service classes so from the cli it needs an UIComponents-free implementation that can be instantiated from cli. In the future this should be a real implementation so the importer could be used from R for example but I would say this is a future feature and only if needed.

File: src/PKSim.Core/PKSim - Backup.Core.csproj (hunk 6 above)
Comment (abdelr) [reply src/PKSim.Core/PKSim - Backup.Core.csproj:1]: Done

File: src/PKSim.Infrastructure/Services/ImportObservedDataTask.cs [hunk 10]
```diff
@@ -1,19 +1,27 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Xml.Linq;
+using FluentNHibernate.Conventions;
```
[the remark this answers is not in the document]
Comment (abdelr) [reply src/PKSim.Infrastructure/Services/ImportObservedDataTask.cs:5]: Done

File: src/PKSim.Infrastructure/Services/ImportObservedDataTask.cs [hunk 11]
```diff
@@ -43,16 +55,71 @@ public class ImportObservedDataTask : IImportObservedDataTask, IObservedDataConf
...
       public void AddObservedDataToProjectForCompound(Compound compound)
       {
          _executionContext.Load(compound);
          addObservedData(importConfiguration, compound);
       }
 
+      public void AddObservedDataFromXmlToProjectForCompound(Compound compound)
+      {
+         _executionContext.Load(compound);
+         AddObservedDataFromXml(importConfiguration, compound);
```
[the remark this answers is not in the document]
Comment (abdelr) [reply src/PKSim.Infrastructure/Services/ImportObservedDataTask.cs:79]: If you blame the file, you will see that importConfiguration comes from you (commit 5580a618) and it is a valid PKSim configuration for the importer. Yes, it does compile.

File: src/PKSim.Presentation/Presenters/ContextMenus/ObservedDataFolderContextMenu.cs [hunk 12]
```diff
@@ -43,6 +43,12 @@ public class ObservedDataFolderContextMenu : ContextMenu
          if (allCompounds.Any())
             _view.AddMenuItem(addObservedDataFor);
 
+         //AddObservedDataFromConfiguration
+         _view.AddMenuItem(repository[MenuBarItemIds.AddObservedDataFromConfiguration]);
+       //  if (allCompounds.Any())
```
[the remark this answers is not in the document]
Comment (abdelr) [reply src/PKSim.Presentation/Presenters/ContextMenus/ObservedDataFolderContextMenu.cs:48]: Done

File: src/PKSim.Infrastructure/Reporting/TEX/Reporters/ProjectReporter.cs (hunk 9 above)
Comment (abdelr) [reply src/PKSim.Infrastructure/Reporting/TEX/Reporters/ProjectReporter.cs]: Done

File: src/PKSim.Presentation/UICommands/ReloadAllObservedDataCommand.cs [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+         )
+      {
+         _executionContext = executionContext;
+         _importObservedDataTask = importObservedDataTask;
+         _observedDataTask = observedDataTask;
+      }
+
+      protected override void PerformExecute()
+      {
+         if (!Subject.ExtendedProperties.Contains("Configuration"))
```
[the remark this answers is not in the document]
Comment (abdelr) [reply src/PKSim.Presentation/UICommands/ReloadAllObservedDataCommand.cs:28]: I agree, we should have a member for this.

[the remark this answers is not in the document]
Comment (abdelr) [reply src/PKSim.Presentation/UICommands/ReloadAllObservedDataCommand.cs]: I moved it into a DataRepository

File: src/PKSim.Presentation/UICommands/ReloadObservedDataCommand.cs [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+      protected override void PerformExecute()
+      {
+         if (!Subject.ExtendedProperties.Contains("Configuration"))
+            return;
+
+         var project = _executionContext.Project;
+         _observedDataTask.Delete(Subject);
+
+         var configuration = project.ImporterConfigurationBy(Subject.ExtendedPropertyValueFor("Configuration"));
```
[the remark this answers is not in the document]
Comment (abdelr) [reply src/PKSim.Presentation/UICommands/ReloadObservedDataCommand.cs]: Done

File: src/PKSim.Infrastructure/Services/ImportObservedDataTask.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (msevestre) [reply src/PKSim.Infrastructure/Services/ImportObservedDataTask.cs:78]: From the diff only, I do not see where this parameter is defined. I will have to load the code in PKSim to review.
And I really don't care who writes the code. If it doesn't make sense, it should change

[the remark this answers is not in the document]
Comment (abdelr) [reply src/PKSim.Infrastructure/Services/ImportObservedDataTask.cs:78]: Well, I did not know you were only looking at the diff only. "importConfiguration" is a method on ImportObservedDataTask class and used in some other places. The method configures the columnInfos and yes, it might be better to rename it to columnInfos. Still, from the diff you will have the same doubt meaning, where is it defined since it looks like a local variable but it is actually a private method. I do not recommend changing more than the name of the method.

**Comment by msevestre** [comment]
@abdelr @georgeDaskalaki
What's the status here? The build is still failing. Can you please fix soon so that we can merge and continue working on the importer integration?

**Comment by abdelr** [comment]
> @abdelr @georgeDaskalaki
> What's the status here? The build is still failing. Can you please fix soon so that we can merge and continue working on the importer integration?

@msevestre I believe it is failing because we are using a prerelease version of core. So I guess when we merge the corresponding branch of core into development and we manage to have a released dll we should be able to build but it is only my guess. We could talk about it tomorrow.

File: src/PKSim.Core/CoreConstants.cs [hunk 15]
```diff
@@ -1335,9 +1335,9 @@ public static class ObservedData
 
          public static IReadOnlyList<string> DefaultProperties = new List<string>
          {
-            OSPSuite.Assets.ObservedData.ORGAN,
-            OSPSuite.Assets.ObservedData.COMPARTMENT,
-            OSPSuite.Assets.ObservedData.MOLECULE,
+            OSPSuite.Assets.ObservedData.Organ,
```
Comment (msevestre) [inline src/PKSim.Core/CoreConstants.cs:1338]: @abdelr @georgeDaskalakis 
Why was this renamed? Constants are normally ALL_CAPS. Let's make sure we revert this in the next update in core

**Review by msevestre — COMMENTED** [review]
Goign to merge
Some TODOS left in the ImportObservedDataTask
